### ISSUE TITLE
Unable to Connect to Mi gateway

### ISSUE THREAD

**Issue by shatulsky** [sergey-brutsky/mi-home#15]
Hi. I tried your code example from `MiHomeConsole`. And it is not working for me.
`miHome.GetGateway()` is always null and `miHome.GetDevices()` is always empty. I tried both ways - with password and SSID and without them.

Here is my HUB info
![photo_2020-02-15_00-32-53 (2)](https://user-images.githubusercontent.com/22248285/74574088-2e0eed00-4f8b-11ea-8db6-3dc840ca6fe4.jpg)
![photo_2020-02-15_00-32-53](https://user-images.githubusercontent.com/22248285/74574096-3109dd80-4f8b-11ea-8e35-fbcb83789395.jpg)

I tried to connect to router via Ethernet cable and via Wi-Fi.
IPTV router setting is enabled:
![image](https://user-images.githubusercontent.com/22248285/74574148-5dbdf500-4f8b-11ea-9bfb-70f1827cf3de.png)

I do not have Hyper-V/Virtual box network adapters:
![image](https://user-images.githubusercontent.com/22248285/74574174-75957900-4f8b-11ea-90dd-08f92da8e8ee.png)

My firewall and antivirus are disabled.
I created 4 firewall rules to allow TCP and UDP ingoing and outgoing connection on 9898 port.

I created the forwarding rule for my router:
![image](https://user-images.githubusercontent.com/22248285/74574304-eb014980-4f8b-11ea-8163-ba9ead850f31.png)

Domoticz is also not working for me. It stuck on `http://127.0.0.1:8080/#/Devices` page:
![image](https://user-images.githubusercontent.com/22248285/74574488-82669c80-4f8c-11ea-879d-5a2369d9aabc.png)

Have you any more suggestions?
Thank you in advance!

**Comment by sergey-brutsky** [comment]
Could you please show me your code that is supposed to b3 working ?

**Comment by shatulsky** [comment]
> Could you please show me your code that is supposed to b3 working ?

```
using MiHomeLib;
using System;
using System.Linq;

namespace MiHomeServer
{
    class Program
    {
        static void Main(string[] args)
        {
            bool pause = false;

            using (var miHome = new MiHome("233c0hrpctwp5sxt"))
            {
                var devices = miHome.GetDevices();
                if (devices.Any())
                {
                    pause = true;
                    foreach (var miHomeDevice in miHome.GetDevices())
                    {
                        Console.WriteLine($"{miHomeDevice.Sid}, {miHomeDevice.GetType()}, {miHomeDevice}");
                    }
                }

                if (miHome.GetGateway() != null)
                {
                    pause = true;
                    Console.WriteLine($"{miHome.GetGateway()?.Sid}");
                }
            }

            if (pause)
            {
                Console.ReadLine();
            }
        }
    }
}
```

**Comment by sergey-brutsky** [comment]
Is your GW and PC where you running your code is the same LAN subnet (without any routers/firewalls/etc) ? 
I'm asking this because it's not clear why do you configuring your FW rules

**Comment by shatulsky** [comment]
> Is your GW and PC where you running your code in the same LAN subnet (without any routers/firewalls/etc) ?

GW is connected to the wi-fi network of my router. PC wich I used to test the code, I sent in the previous message, is also connected to the same router via Ethernet cable.

I also tried that example code on the notebook connected to the same wi-fi hotspot (with the gateway) and result is the same.

> I'm asking this because it's not clear why do you configuring your FW rules
I just tries all possible solution that can help in my opinion.

**Comment by sergey-brutsky** [comment]
Could you please add a small pause after `using` and before `GetDevices` method like `Thread.Sleep` or `Task.Delay` for 5 seconds and check if any devices obtained from your GW ?

**Comment by shatulsky** [comment]
> Could you please add a small pause after `using` and before `GetDevices` method like `Thread.Sleep` or `Task.Delay` for 5 seconds and check if any devices obtained from your GW ?
Tested it on my PC connected to the same with Gateway router via Ethernet cable.
Deley did not help. `miHome.GetDevices()` is still empty collection. And `miHome.GetGateway()` is still null. :(

**Comment by sergey-brutsky** [comment]
Do you use any technologies like VirtualBox or VPN that may block UDP multicast traffic on your PC ? I saw similar symptoms in another issue.

**Comment by sergey-brutsky** [comment]
Also please try to run this simple python script on some device inside your LAN
```
import struct
from socket import *

print("Start listening socket ...")
s = socket(AF_INET, SOCK_DGRAM, SOL_UDP)
s.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
s.setsockopt(IPPROTO_IP, IP_ADD_MEMBERSHIP, struct.pack('4sL', inet_aton('224.0.0.50'), INADDR_ANY))
s.bind(('', 9898))
s.sendto('{\"cmd\":\"get_id_list\"}'.encode(),  ('224.0.0.50', 9898))

while True:
	data, x = s.recvfrom(2046)
	print(x, data)

```
You should see that data are going back and forth, if they not it means that something is blocking your UDP multicasts. 
What it could be ? This is a very interesting question. 
From my experience it could be everything starting from your OS network settings till router/firewall.

**Comment by shatulsky** [comment]
> Do you use any technologies like VirtualBox or VPN that may block UDP multicast traffic on your PC ? I saw similar symptoms in another issue [#7 (comment)](https://github.com/sergey-brutsky/mi-home/issues/7#issuecomment-424507293)

I had only this driver 
![image](https://user-images.githubusercontent.com/22248285/74602362-53e1e200-50b0-11ea-8fce-835601be79a6.png)
But it is already disabled.
Have no other ideas about what can block the traffic.

**Comment by shatulsky** [comment]
> Also please try to run this simple python script on some device inside your LAN

I tried it on my PC and Notebook.
When the script is started on PC and I start it on Notebook - 1 more log line appends on PC (last 3 lines).
![image](https://user-images.githubusercontent.com/22248285/74602620-4bd77180-50b3-11ea-8ccd-1827ef796351.png)

When I start the new script instance on PC with an already launched instance on Notebook - log with 192.168.0.101 appends.
![image](https://user-images.githubusercontent.com/22248285/74602696-3282f500-50b4-11ea-9ce4-649120ff9d3c.png)

So, as for me, result of this investigation are:
- PC is on 192.168.0.101
- Notebook is on 192.168.0.103
- The UDP multicast connection is not blocked by anything.

**Comment by sergey-brutsky** [comment]
This logs shows that UDP traffic is blocked by something because you don't get any response

**Comment by sergey-brutsky** [comment]
This is how NOT blocked UDP multicast should looks like
```
Start listening socket ...
(('192.168.1.14', 9898), '{"cmd":"get_id_list"}')
(('192.168.1.12', 9898), '{"cmd":"get_id_list_ack","sid":"34ce0088db36","token":"mWnYERoTKhukKxKQ","data":"[\\"158d0001826509\\",\\"158d00015dc662\\",\\"158d00015dc6cc\\",\\"158d000182e561\\",\\"158d000182e3aa\\",\\"158d0001826527\\",\\"158d000182dfbc\\",\\"158d000187437b\\",\\"158d00011c0754\\",\\"158d000156a94b\\",\\"158d0001874ad8\\",\\"158d00015a9b17\\",\\"158d0001214e1f\\",\\"158d00015da18e\\",\\"158d0001255529\\",\\"158d0001255565\\",\\"158d00012554c0\\",\\"158d0001116fec\\",\\"158d0001d545d0\\",\\"158d0001d593e2\\",\\"158d0001db9340\\",\\"158d0001d8f839\\",\\"158d0001d948a0\\",\\"158d0001d8f8f7\\"]"}')
(('192.168.1.12', 4321), '{"cmd":"heartbeat","model":"gateway","sid":"34ce0088db36","short_id":"0","token":"gnj2xHQfFHq6LjlG","data":"{\\"ip\\":\\"192.168.1.12\\"}"}')
(('192.168.1.12', 4321), '{"cmd":"report","model":"motion","sid":"158d0001214e1f","short_id":19149,"data":"{\\"status\\":\\"motion\\"}"}')
```
There are requests and responses here from GW here
In your case something is blocking UDP responses from GW, probably non-standard OS network stack configuration

**Comment by sergey-brutsky** [comment]
@shatulsky could you please temporary uninstall VirtualBox NDIS driver (reboot) and double check UDP traffic again ?

**Comment by shatulsky** [comment]
> @shatulsky could you please temporary uninstall VirtualBox NDIS driver (reboot) and double check UDP traffic again ?

The results are the same.

**Comment by shatulsky** [comment]
I looked at Wireshark report and I'm actually receiving some responses from 192.168.0.102 device
Maybe it will help you
![image](https://user-images.githubusercontent.com/22248285/74613548-0bf6a580-5118-11ea-9cbb-162d102ad319.png)

**Comment by sergey-brutsky** [comment]
@shatulsky could you please change multicast addr in the python script above from 224.0.0.50 to 224.0.0.51 and run it again ?
Maybe in newer version of GW or firmware they changed multicast address.

**Comment by shatulsky** [comment]
> 224.0.0.51

It did not help.
What about router settings? As for now, I have
![image](https://user-images.githubusercontent.com/22248285/74634390-6faabe00-516c-11ea-8bc9-fa1cf4f320b7.png)
![image](https://user-images.githubusercontent.com/22248285/74634413-79ccbc80-516c-11ea-8528-433085a2a9c2.png)
![image](https://user-images.githubusercontent.com/22248285/74634461-923cd700-516c-11ea-8b10-a5bc11dcba8c.png)

The router supports 2 more modes for IPTV
![image](https://user-images.githubusercontent.com/22248285/74634522-b00a3c00-516c-11ea-83e4-93f847d395bf.png)
![image](https://user-images.githubusercontent.com/22248285/74634536-b7314a00-516c-11ea-91f0-003dbda5f681.png)

**Comment by sergey-brutsky** [comment]
Is it possible to stop firewall on your router ?

**Comment by shatulsky** [comment]
> Is it possible to stop firewall on your router ?

It was already disabled before I created this issue. I double-checked it, still disabled.

Have you any example of code that will simulate the multicast server. I can try to test this code on the notebook connected to the Wi-Fi hotspot and listen to it using python script, you already sent, from PC. It will help to determine on what side the problem is, network configuration or HUB broadcasting

**Comment by pachman** [comment]
Hello guys! I have the same problem.
After I turn off "VirtualBox Host-Only Network", the python script got useful information

`('192.168.1.36', 9898) b'{"cmd":"get_id_list"}'
('192.168.1.52', 9898) b'{"cmd":"get_id_list_ack","sid":"04cf8c982f33","token":"","data":"[\\"158d00031b32c1\\",\\"158d00032170b8\\",\\"158d0002ec3079\\",\\"158d00028999e8\\"]"}'
('192.168.1.52', 4321) b'{"cmd":"heartbeat","model":"gateway","sid":"04cf8c982f33","short_id":"0","token":"","data":"{\\"ip\\":\\"192.168.1.52\\"}"}'
('192.168.1.52', 4321) b'{"cmd":"heartbeat","model":"gateway","sid":"04cf8c982f33","short_id":"0","token":"","data":"{\\"ip\\":\\"192.168.1.52\\"}"}'
('192.168.1.36', 9898) b'{"cmd":"get_id_list"}'
('192.168.1.52', 9898) b'{"cmd":"get_id_list_ack","sid":"04cf8c982f33","token":"","data":"[\\"158d00031b32c1\\",\\"158d00032170b8\\",\\"158d0002ec3079\\",\\"158d00028999e8\\"]"}'`

but from dotnet code I get nothing

**Comment by sergey-brutsky** [comment]
@pachman could you please specify what version of .net runtime do you use ? 
As far as I remember on .net core 2.1 there were some issues with UDP multicast traffic

**Comment by sergey-brutsky** [comment]
@shatulsky could you please provide me with the information what .net runtime do you use for your project and OS ?

**Comment by pachman** [comment]
> @pachman could you please specify what version of .net runtime do you use ?
> As far as I remember on .net core 2.1 there were some issues with UDP multicast traffic

I am using the latest version 3.1

**Comment by shatulsky** [comment]
> @shatulsky could you please provide me with the information what .net runtime do you use for your project and OS ?

The console app is on .net framework 4.7.2.
The installed version is .net framework 4.8

**Comment by sergey-brutsky** [comment]
@shatulsky I will try to run your piece of code on the framework mentioned above against my setup router + gw and let you know if it is working

**Comment by sergey-brutsky** [comment]
> > @pachman could you please specify what version of .net runtime do you use ?
> > As far as I remember on .net core 2.1 there were some issues with UDP multicast traffic
> 
> I am using the latest version 3.1

Could you please paste here non-working code ?

**Comment by fadidevv** [comment]
Hi,

your library works with Token Aqara Gateway method or is i need to enable LAN MODE from MI app to get the Password? As you know MI app does not allow anymore to enable LAN MODE from the APP so cannot get password but this library works with UDP mode? Is my Aqara have to connected in my WIFI network to use this library? I test with my Aqara Gateway latest version hardware with wireshark and UDP i can copy but when i use Packet sender to send command via UDP nothing happens is it mean my network have no UDP allow?

Thank you.

**Comment by sergey-brutsky** [comment]
hi @fadidevv 

1) This is not necessary, password is required only when you want to "send" commands to your mi home devices like "switch light on" or "turn radio on gw". If you just "read" data from your sensors, password is not needed.

2) Your application using library MUST run on the same lan segment where your GW is.
For example if your GW has ip/netmask 192.168.1.1/255, your app must run in the segment 192.168.1.x/255

3) Not clear what to you mean by "when i use Packet sender to send command via UDP nothing happens". Could you please elaborate a little bit ?

**Comment by fadidevv** [comment]
Thanks for the reply, i see your library use Multicast Traffic method like LAN MODE and you said in docs have to open MI Home app to enable the LAN mode? i tested your library in Aqara HUB v3 but not working i'm on wifi same IP of device and my computer, i try to use UDP method by using wireshark but when i send request with HEX packet nothing shows in devices nothing returns!

**Comment by sergey-brutsky** [comment]
@fadidevv could you please at first run python script mentioned above in order to check that UPD multicast traffic is working in your lan segment ?

**Comment by fadidevv** [comment]
> @fadidevv could you please at first run python script mentioned above in order to check that UPD multicast traffic is working in your lan segment ?

Hi,

<img width="559" alt="Screen Shot 2020-02-21 at 3 40 40 PM" src="https://user-images.githubusercontent.com/43948953/75017908-934d6b80-54c0-11ea-9816-1ad83b5e1439.png">

Nothing returns, my device gateway is paired with my WIFI HOTSPOT in the same network from where i ran this python script.

**Comment by fadidevv** [comment]
And here is the Wireshark logs same as @shatulsky Looks like Aqara Hub Gateway UDP ip has changed take a look @sergey-brutsky 

<img width="1983" alt="Screen Shot 2020-02-21 at 3 45 22 PM" src="https://user-images.githubusercontent.com/43948953/75018356-52098b80-54c1-11ea-87ea-17cbba095d72.png">

**Comment by sergey-brutsky** [comment]
@shatulsky @fadidevv could you please check that required ports are open on your GW version
`nmap -sU <ip of your gw here> -p 9898,4321,54321`

**Comment by shatulsky** [comment]
> `nmap -sU <ip of your gw here> -p 9898,4321,54321`
```
Starting Nmap 7.80 ( https://nmap.org ) at 2020-02-22 00:02 Oeieyiaey (ceia)
Nmap scan report for 192.168.0.102
Host is up (0.087s latency).

PORT      STATE         SERVICE
4321/udp  closed        rwhois
9898/udp  closed        monkeycom
54321/udp open|filtered bo2k
MAC Address: 04:CF:8C:F6:7C:B5 (Xiaomi Electronics,co.)

Nmap done: 1 IP address (1 host up) scanned in 2.09 seconds
```

**Comment by sergey-brutsky** [comment]
@shatulsky @fadidevv it looks like that manufacturer on some versions of GW closed ports required for normal UDP multicast working. 
Please check this issue --> https://community.openhab.org/t/solved-openhab2-xiaomi-mi-gateway-does-not-respond/52963/107

I found an instruction how to open required ports --> https://community.openhab.org/t/solved-openhab2-xiaomi-mi-gateway-does-not-respond/52963/114

Could you guys check if the library working after opening ports by using the instructions above ?
Thanks

**Comment by sergey-brutsky** [comment]
@shatulsky @fadidevv guys, did you have a chance to check if the method mentioned above is working ?

**Comment by shatulsky** [comment]
> @shatulsky @fadidevv guys, did you have a chance to check if the method mentioned above is working ?

I tried the suggested method. But my putty output is empty.
I'm going to make a few more attempts soon. I'll keep in touch.

**Comment by fadidevv** [comment]
@sergey-brutsky Nothing returns everything is empty, as i understand the new hardware models from Aqara don't work anymore with local-LAN networks i apply for Aqara Cloud and they approve my account and can only use Aqara Cloud APIS via HTTP, LAN mode only works with some Aqara devices.

I'm going to reverse the Aqara or MI Home APP and bypass the security like signatures-random in request headers and i will make HTTP APIS for Aqara.

**Comment by sergey-brutsky** [comment]
@fadidevv what this command `nmap -sU <ip of your gw here> -p 9898,4321,54321` output ?

**Comment by shatulsky** [comment]
> @shatulsky @fadidevv guys, did you have a chance to check if the method mentioned above is working ?

I successfully changed network.open_pf property on Gateway using [this instruction](https://community.openhab.org/t/solved-openhab2-xiaomi-mi-gateway-does-not-respond/52963/114). I have one more clarification that may help someone in the future. On the screenshot attached to the instruction RX and TX are the gateway connectors but not UART cable. So you should connect UART RX to TX on the screenshot and UART TX to TX on the screenshot.

After that gateway opened its ports.
```
nmap -sU 192.168.0.102 -p 9898,4321,54321
4321/udp  open|filtered rwhois
9898/udp  open          monkeycom
54321/udp open|filtered bo2k
```

Python script receives some data -
```
Start listening socket ...
('192.168.0.103', 9898) b'{"cmd":"get_id_list"}'
('192.168.0.102', 9898) b'{"cmd":"get_id_list_ack","sid":"04cf8cf67cb5","token":"adRtLQrrjS2QrO7m","data":"[\\"158d00034de7ed\\"]"}'
('192.168.0.102', 4321) b'{"cmd":"heartbeat","model":"gateway","sid":"04cf8cf67cb5","short_id":"0","token":"Kw8DMSRNp2oRsTkl","data":"{\\"ip\\":\\"192.168.0.102\\"}"}'
